flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/MatchCodeGenerator.scala

```diff
@@ -0,0 +1,383 @@
+    if (fieldRef.getAlpha.equals("*") && currentPattern.isDefined) {
+      generateFieldAccess(input, input1Term, fieldRef.getIndex)
+    } else {
+      generatePatternFieldRef(fieldRef)
+    }
+  }
+
+  private def generateDefinePatternVariableExp(
+      patternName: String,
+      currentPattern: String)
+    : GeneratedPatternList = {
+    val listName = newName("patternEvents")
+    val eventTypeTerm = boxedTypeTermForTypeInfo(input)
+    val eventNameTerm = newName("event")
+
+    val addCurrent = if (currentPattern == patternName) {
+      s"""
+         |$listName.add($input1Term);
+         """.stripMargin
+    } else {
+      ""
+    }
+
+    val escapedPatternName = StringEscapeUtils.escapeJava(patternName)
+    val code =
+      s"""
+         |java.util.List $listName = new java.util.ArrayList();
+         |for ($eventTypeTerm $eventNameTerm :
+         |  $contextTerm.getEventsForPattern("$escapedPatternName")) {
+         |    $listName.add($eventNameTerm);
+         |}
+         |$addCurrent
+         |""".stripMargin
+
+    GeneratedPatternList(s"$listName", code)
+  }
+
+  private def generateMeasurePatternVariableExp(patternName: String): GeneratedPatternList = {
+    val listName = newName("patternEvents")
+
+    val code = if (patternName == "*") {
+      s"""
+         |java.util.List $listName = new java.util.ArrayList();
+         |for (java.util.List entry : $input1Term.values()) {
+         |  $listName.addAll(entry);
+         |}
+         """.stripMargin
+    } else {
+      val escapedPatternName = StringEscapeUtils.escapeJava(patternName)
+      s"""
+         |java.util.List $listName = (java.util.List) $input1Term.get("$escapedPatternName");
+         |if ($listName == null) {
+         |  $listName = java.util.Collections.emptyList();
+         |}
+         |""".stripMargin
+    }
+
+    GeneratedPatternList(s"$listName", code)
+  }
+
+  private def findEventByLogicalPosition(
+      patternFieldAlpha: String)
+    : GeneratedExpression = {
+    val rowNameTerm = newName("row")
+    val eventTypeTerm = boxedTypeTermForTypeInfo(input)
+    val isRowNull = newName("isRowNull")
+
+    val findEventsByPatternName = reusablePatternLists.get(patternFieldAlpha) match {
+      // input access and unboxing has already been generated
+      case Some(expr) =>
+        expr
+
+      case None =>
+        val exp = currentPattern match {
+          case Some(p) => generateDefinePatternVariableExp(patternFieldAlpha, p)
+          case None => generateMeasurePatternVariableExp(patternFieldAlpha)
+        }
+        reusablePatternLists(patternFieldAlpha) = exp
+        exp
+    }
+
+    val listName = findEventsByPatternName.resultTerm
+    val resultIndex = if (first) {
+      s"""$offset"""
+    } else {
+      s"""$listName.size() - $offset - 1"""
+    }
+
+    val funcCode =
+      s"""
+         |$eventTypeTerm $rowNameTerm = null;
+         |boolean $isRowNull = true;
```

Comment:
Right now this boolean flag is not needed as we access the result with a nullable input field access. Do we need to call `generateNullableInputFieldAccess` or can we use `generateFieldAccess`?